esp-radio/src/wifi/net.rs

```diff
@@ -0,0 +1,280 @@
+        dump_packet_info(buffer);
+
+        f(buffer)
+    }
+}
+
+#[doc(hidden)]
+#[derive(Debug)]
+pub struct WifiTxToken {
+    mode: WifiDeviceMode,
+}
+
+impl WifiTxToken {
+    /// Consumes the TX token and applies the callback function to the received
+    /// data buffer.
+    fn consume_token<R, F>(self, len: usize, f: F) -> R
+    where
+        F: FnOnce(&mut [u8]) -> R,
+    {
+        self.mode.increase_in_flight_counter();
+
+        // (safety): creation of multiple Wi-Fi devices with the same mode is impossible
+        // in safe Rust, therefore only smoltcp _or_ embassy-net can be used at
+        // one time
+        static mut BUFFER: [u8; MTU] = [0u8; MTU];
+
+        let buffer = unsafe { &mut BUFFER[..len] };
+
+        let res = f(buffer);
+
+        esp_wifi_send_data(self.mode.interface(), buffer);
+
+        res
+    }
+}
+
+impl WifiDeviceMode {
+    fn tx_token(&self) -> Option<WifiTxToken> {
+        if !self.can_send() {
+            // TODO: perhaps we can use a counting semaphore with a short blocking timeout
+            crate::preempt::yield_task();
+        }
+
+        if self.can_send() {
+            Some(WifiTxToken { mode: *self })
+        } else {
+            None
+        }
+    }
+
+    fn rx_token(&self) -> Option<(WifiRxToken, WifiTxToken)> {
+        let is_empty = self.data_queue_rx().with(|q| q.is_empty());
+        if is_empty || !self.can_send() {
+            // TODO: use an OS queue with a short timeout
+            crate::preempt::yield_task();
+        }
+
+        let is_empty = is_empty && self.data_queue_rx().with(|q| q.is_empty());
+
+        if !is_empty {
+            self.tx_token().map(|tx| (WifiRxToken { mode: *self }, tx))
+        } else {
+            None
+        }
+    }
+}
+
+/// The embassy-net driver implementation.
+pub mod embassy {
+    use embassy_net_driver::{Capabilities, Driver, HardwareAddress, RxToken, TxToken};
+    use esp_hal::asynch::AtomicWaker;
+
+    use super::*;
+
+    // We can get away with a single tx waker because the transmit queue is shared
+    // between interfaces.
+    pub(crate) static TRANSMIT_WAKER: AtomicWaker = AtomicWaker::new();
+
+    pub(crate) static AP_RECEIVE_WAKER: AtomicWaker = AtomicWaker::new();
+    pub(crate) static AP_LINK_STATE_WAKER: AtomicWaker = AtomicWaker::new();
+
+    pub(crate) static STA_RECEIVE_WAKER: AtomicWaker = AtomicWaker::new();
+    pub(crate) static STA_LINK_STATE_WAKER: AtomicWaker = AtomicWaker::new();
+
+    impl RxToken for WifiRxToken {
+        fn consume<R, F>(self, f: F) -> R
+        where
+            F: FnOnce(&mut [u8]) -> R,
+        {
+            self.consume_token(f)
+        }
+    }
+
+    impl TxToken for WifiTxToken {
+        fn consume<R, F>(self, len: usize, f: F) -> R
+        where
+            F: FnOnce(&mut [u8]) -> R,
+        {
+            self.consume_token(len, f)
+        }
+    }
+
+    /// A [Driver] implementation.
+    pub struct EmbassyNetAdapter<'d> {
```

Comment:
(I can still drop the WifiController while this is alive, I think.)

Are these adapter structs really something we want/need? It's a really unfortunate thing the users would need to interact with. This is necessary if users wanted to support another network stack, but for the ones we implement ourselves, why can't we just implement the traits for WifiDevice?